File: EULER/38_pandigital_multiples.py
from itertools import permutations

def calc_pandigital(n: int):
  ans, multiplier = "", 0
  while len(ans) < 9:
    multiplier += 1
    ans += str(n * multiplier)
    if len(ans) >= 9:
      return ans if sorted(ans) == sorted("123456789") else 0

def is_achievable(pandigital: str):
  for n_width in range(1, 5):
    seed = int(pandigital[:n_width])
    if calc_pandigital(seed) == pandigital:
      return True
  return False

for pandigital_candidate in permutations("987654321"):
  pandigital_candidate = "".join(pandigital_candidate)
  if is_achievable(pandigital_candidate):
    print(pandigital_candidate)
    break

File: EULER/38_pandigital_multiples.cpp
#include <bits/stdc++.h>
using namespace std;

string calc_pandigital(int n){
  string ans = "";
  int multiplier = 0;
  while (ans.length() < 9) {
    multiplier += 1;
    ans += to_string(n * multiplier);
    if (ans.length() >= 9){
      string ans_copy = ans;
      sort(ans_copy.begin(), ans_copy.end());
      if (ans_copy == "123456789") return ans;
      return "0";
    }
  }
  return "0";
}

bool is_achievable(string pandigital){
  for (int n_width=1; n_width<6; n_width++){
    int seed = stoi(pandigital.substr(0, n_width));
    if (calc_pandigital(seed) == pandigital) return true;
  }
  return false;
}

int main(){
  string pandigital_candidate = "987654321";
  do {
    if (is_achievable(pandigital_candidate)){
      cout << pandigital_candidate << endl;
      break;
    }
  } while (prev_permutation(pandigital_candidate.begin(), pandigital_candidate.end()));
}
